process/smartContract/processorV2/processV2.go

```diff
@@ -0,0 +1,2645 @@
+		if err != nil {
+			log.Trace("scProcessor.createSCRsWhenError()", "error parsing args", data)
+			return "", err
+		}
+
+		if len(args) < 2 {
+			log.Trace("scProcessor.createSCRsWhenError()", "no async params found", data)
+			return "", err
+		}
+		data = function
+		newArgs = append(args, newAsyncParams...)
+	} else {
+		newArgs = newAsyncParams
+	}
+
+	for _, arg := range newArgs {
+		data += "@" + hex.EncodeToString(arg)
+	}
+
+	return data, nil
+}
+
+func mergeVMOutputLogs(newVMOutput *vmcommon.VMOutput, vmOutput *vmcommon.VMOutput) {
+	if len(vmOutput.Logs) == 0 {
+		return
+	}
+
+	if newVMOutput.Logs == nil {
+		newVMOutput.Logs = make([]*vmcommon.LogEntry, 0, len(vmOutput.Logs))
+	}
+
+	newVMOutput.Logs = append(vmOutput.Logs, newVMOutput.Logs...)
+}
+
+func (sc *scProcessor) resolveBuiltInFunctions(
+	vmInput *vmcommon.ContractCallInput,
+) (*vmcommon.VMOutput, error) {
+
+	vmOutput, err := sc.blockChainHook.ProcessBuiltInFunction(vmInput)
+	if err != nil {
+		vmOutput = &vmcommon.VMOutput{
+			ReturnCode:    vmcommon.UserError,
+			ReturnMessage: err.Error(),
+			GasRemaining:  0,
+		}
+
+		return vmOutput, nil
+	}
+
+	return vmOutput, nil
+}
+
+func (sc *scProcessor) isSameShardSCExecutionAfterBuiltInFunc(
+	tx data.TransactionHandler,
+	vmInput *vmcommon.ContractCallInput,
+	vmOutput *vmcommon.VMOutput,
+) (bool, vmData.CallType, *vmcommon.ParsedESDTTransfers, *vmcommon.OutputTransfer, error) {
+	if vmOutput.ReturnCode != vmcommon.Ok {
+		return false, 0, nil, nil, nil
+	}
+
+	parsedTransfer, err := sc.esdtTransferParser.ParseESDTTransfers(vmInput.CallerAddr, vmInput.RecipientAddr, vmInput.Function, vmInput.Arguments)
+	if err != nil {
+		return false, 0, nil, nil, nil
+	}
+	if !core.IsSmartContractAddress(parsedTransfer.RcvAddr) {
+		return false, 0, nil, nil, nil
+	}
+
+	if sc.shardCoordinator.ComputeId(parsedTransfer.RcvAddr) != sc.shardCoordinator.SelfId() {
+		return false, 0, nil, nil, nil
+	}
+
+	callType := determineCallType(tx)
+	if callType == vmData.AsynchronousCallBack {
+		return true, callType, parsedTransfer, nil, nil
+	}
+
+	if len(parsedTransfer.CallFunction) == 0 {
+		return false, 0, nil, nil, nil
+	}
+
+	outAcc, ok := vmOutput.OutputAccounts[string(parsedTransfer.RcvAddr)]
+	if !ok {
+		return false, 0, nil, nil, nil
+	}
+	if len(outAcc.OutputTransfers) != 1 {
+		return false, 0, nil, nil, nil
+	}
+
+	scExecuteOutTransfer := outAcc.OutputTransfers[0]
+	return true, callType, parsedTransfer, &scExecuteOutTransfer, nil
+}
+
+func (sc *scProcessor) prepareExecutionAfterBuiltInFunc(
+	callType vmData.CallType,
+	vmInput *vmcommon.ContractCallInput,
+	vmOutput *vmcommon.VMOutput,
+	parsedTransfer *vmcommon.ParsedESDTTransfers,
+	scExecuteOutTransfer *vmcommon.OutputTransfer,
+	tx data.TransactionHandler,
+	acntSnd state.UserAccountHandler,
+	snapshot int) (*vmcommon.ContractCallInput, state.UserAccountHandler, error) {
+
+	var newVMInput *vmcommon.ContractCallInput
+	if callType == vmData.AsynchronousCallBack {
```

Comment:
in the original code we do not have this if branch

Reply:
The logic exists in function isSCExecutionAfterBuiltInFunc(). Unfortunately this function did too many things besides answering the question that it's name implies.